### PR TITLE
Add degree checks.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: std/machines/binary.asm [hunk 1]
```diff
@@ -8,9 +8,6 @@ machine Binary with
     // Allow this machine to be connected via a permutation
     call_selectors: sel,
 {
-
-    // lower bound degree is 262144
```
Comment (chriseth) [inline std/machines/binary.asm:12]: I'm counting 196608 - am I doing something wrong?

Comment (leonardoalt) [reply std/machines/binary.asm:12]: This one I really thought was 2**18, but yea also don't see why. Maybe next power of 2

File: std/array.asm [hunk 2]
```diff
@@ -29,6 +29,9 @@ let<T1, T2> fold: T1[], T2, (T2, T1 -> T2) -> T2 = |arr, initial, folder| std::u
 // use this function to flatten an array of arrays.
 let<T: Add + FromLiteral> sum: T[] -> T = |arr| fold(arr, 0, |a, b| a + b);
 
+/// Returns the product of the array elements.
+let<T: Mul + FromLiteral> product: T[] -> T = |arr| fold(arr, 0, |a, b| a * b);
```
Comment (leonardoalt) [inline std/array.asm:33]:
```suggestion
let<T: Mul + FromLiteral> product: T[] -> T = |arr| fold(arr, 1, |a, b| a * b);
```
?

Comment (chriseth) [reply std/array.asm]: Good one!

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**